File: Programación Orientada a Objetos (PDOO)/Napakalaki(Ruby)/pkg/pruebaNapakalaki-0.0.1/lib/card_dealer.rb
#encoding: utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require "singleton"
require_relative "bad_consequence.rb"
require_relative "prize.rb"
require_relative "treasure_kind.rb"
require_relative "monster.rb"
require_relative "treasure.rb"
require_relative "death_bad_consequence.rb"
require_relative "specific_bad_consequence.rb"
require_relative "numeric_bad_consequence.rb"
require_relative "cultist.rb"
require_relative "cultist_player.rb"
module NapakalakiGame

class CardDealer
  
  include Singleton
  
  @@max_value_treasures=10
  
  attr_accessor :unusedTreasures
  attr_accessor :unusedMonsters
  attr_accessor :usedTreasures
  attr_accessor :usedMonsters
  attr_reader :cardDealer
  attr_accessor :unusedCultists
  
  def initialize
   # @cardDealer= CardDealer.instance
    @unusedTreasures=Array.new
    @unusedMonsters=Array.new
    @usedMonsters=Array.new
    @usedTreasures=Array.new
    @unusedCultists = Array.new
  end
  
  def initTreasureCardDeck
    @unusedTreasures << Treasure.new("Sí mi amo!",4,TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Botas de investigación",3,TreasureKind::SHOES)
    @unusedTreasures << Treasure.new("Capucha de Cthulhu",3,TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("A prueba de babas",2,TreasureKind::ARMOR)
    @unusedTreasures << Treasure.new("Botas de lluvia ácida",1,TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Casco minero",2,TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Ametralladora ACME",4,TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Camiseta de la ETSIIT",1,TreasureKind::ARMOR)
    @unusedTreasures << Treasure.new("Clavo de rail ferroviario",3,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Cuchillo de sushi arcano",2,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Fez alópodo",3,TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Hacha prehistórica",2,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("El aparato del Pr. Tesla",4,TreasureKind::ARMOR)
    @unusedTreasures << Treasure.new("Gaita",4,TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Insecticida",2,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Escopeta de 3 cañones",4,TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Garabato mistico",2,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("La rebeca metálica",2,TreasureKind::ARMOR)
    @unusedTreasures << Treasure.new("Necro-playboycon",3,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Lanzallamas",4,TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Necro-comicón",1,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Necronomicón",5,TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Linterna a 2 manos",3,TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Necro-gnomicón",2,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Necrotelecom",2,TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Porra preternatural",2,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Tentácula de pega",2,TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Zapato deja-amigos",1,TreasureKind::SHOES)
    @unusedTreasures << Treasure.new("Shogulador",1,TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Varita de atizamiento",3,TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Mazo de los antiguos",3,TreasureKind::ONEHAND)
  end
  
  def initMonstersCardDeck
    #Creamos todas las cartas:
    pr1 = Prize.new(3,2)
    badc1 = NumericBadConsequence.new("Pierdes 5 niveles y 3 tesoros visibles", 5, 3, 0)
    @unusedMonsters<< Monster.new("El rey rosado",13,pr1,badc1)

    pr2 = Prize.new(4,1)
    badc2 = SpecificBadConsequence.new("Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta", 0, [TreasureKind::ONEHAND], [TreasureKind::ONEHAND])
    @unusedMonsters<< Monster.new("Demonios de Magaluf",2,pr2,badc2)

    pr3 = Prize.new(2,1)
    badc3 = SpecificBadConsequence.new("Pierdes tu armadura visible y otra oculta", 0, [TreasureKind::ARMOR], [TreasureKind::ARMOR])
    @unusedMonsters<< Monster.new("Byakhees de bonanza",8,pr3,badc3)

    pr4 = Prize.new(1,1)
    badc4 = SpecificBadConsequence.new("Embobados con el lindo primigenio te descartas de tu casco visible", 0, [TreasureKind::HELMET], [])
    @unusedMonsters<< Monster.new("Tenochtitlan",2,pr4,badc4)
    
    pr5 = Prize.new(1,1)
    badc5 = SpecificBadConsequence.new("El primordial bostezo contagioso. Pierdes el calzado visible", 0, [TreasureKind::SHOES], [])
    @unusedMonsters<< Monster.new("El sopor de Dunwich",2,pr5,badc5)

    pr6 = Prize.new(3,1)
    badc6 = NumericBadConsequence.new("Pierdes todos tus tesoros visibles",0,@@max_value_treasures,0)
    @unusedMonsters<< Monster.new("El gorron en el umbral",13,pr6,badc6)

    pr7 = Prize.new(2,1)
    badc7 = SpecificBadConsequence.new("Pierdes la armadura visible", 0, [TreasureKind::ARMOR], [])
    @unusedMonsters<< Monster.new("H.P. Munchcraft",6,pr7,badc7)

    pr8 = Prize.new(1,1)
    badc8 = SpecificBadConsequence.new("Sientes bichos bajo la ropa. Descarta la armadura visible", 0, [TreasureKind::ARMOR], [])
    @unusedMonsters<< Monster.new("Necrófago",13,pr8,badc8)

    pr9 = Prize.new(1,1)
    badc9 = NumericBadConsequence.new("Toses los pulmones y pierdes 2 niveles", 2, 0, 0)
    @unusedMonsters<< Monster.new("Flecher",2,pr9,badc9)

    pr10 = Prize.new(2,1)
    badc10 = DeathBadConsequence.new("Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estás muerto")
    @unusedMonsters<< Monster.new("Los hondos",8,pr10,badc10)

    pr11 = Prize.new(2,1)
    badc11 = NumericBadConsequence.new("Pierdes 2 niveles y 2 tesoros ocultos", 2, 0, 2)
    @unusedMonsters<< Monster.new("Semillas Cthulhu",4,pr11,badc11)

    pr12 = Prize.new(2,1)
    badc12 = NumericBadConsequence.new("Te intentas escaquear. Pierdes una mano visible", 0, 1, 0)
    @unusedMonsters<< Monster.new("Dameargo",1,pr12,badc12)

    pr13 = Prize.new(2,1)
    badc13 = NumericBadConsequence.new("Da mucho asquito. Pierdes 3 niveles", 3, 0, 0)
    @unusedMonsters<< Monster.new("Pollipólipo volante", 3, pr13, badc13)

    pr14 = Prize.new(3,1)
    badc14 = DeathBadConsequence.new("No le hace gracia que pronuncien mal su nombre. Estás muerto")
    @unusedMonsters<< Monster.new("Yskhtihyssg-Goth",14,pr14,badc14)

    pr15 = Prize.new(3,1)
    badc15 = DeathBadConsequence.new("La familia te atrapa. Estas muerto")
    @unusedMonsters<< Monster.new("Familia feliz",1,pr15,badc15)

    pr16 = Prize.new(2,1)
    badc16= SpecificBadConsequence.new("La quinta directiva primaria te obliga a perder 2 niveles y un tesoro de 2 manos",2, [TreasureKind::BOTHHANDS],[])
    @unusedMonsters<< Monster.new("Roboggoth",8,pr16,badc16)

    pr17 = Prize.new(1,1)
    badc17 = SpecificBadConsequence.new("Te asusta en la noche. Pierdes un casco visible.",0,[TreasureKind::HELMET],[])
    @unusedMonsters<< Monster.new("El espia sordo",5,pr17,badc17)

    pr18 = Prize.new(2,1)
    badc18= NumericBadConsequence.new("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles",2,5,0)
    @unusedMonsters << Monster.new("Tongue", 19,pr18,badc18)

    pr19 = Prize.new(2,1)
    badc19= SpecificBadConsequence.new("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos",3,[[TreasureKind::ONEHAND],[TreasureKind::ONEHAND],[TreasureKind::BOTHHANDS]],[])
    @unusedMonsters << Monster.new("Bicefalo",21,pr19,badc19)
    
    #MONSTRUOS CON SECTARIOS
    pr20 = Prize.new(3,1)
    bc20 = SpecificBadConsequence.new("Pierdes 1 mano visible",0,[TreasureKind::ONEHAND],[])
    @unusedMonsters << Monster.newMonster("El mal indecible impronunciable",10,pr20,bc20,-2)
    
    pr21 = Prize.new(2,1)
    bc21 = SpecificBadConsequence.new("Pierdes tus tesoros visibles. Jajaja.", 0, [[TreasureKind::ONEHAND],[TreasureKind::ONEHAND],[TreasureKind::BOTHHANDS],[TreasureKind::HELMET],[TreasureKind::ARMOR],[TreasureKind::SHOES]], [])
    @unusedMonsters << Monster.newMonster("Testigos Coulares",6,pr21,bc21,2)
    
    pr22 = Prize.new(2,5)
    bc22= DeathBadConsequence.new("Hoy no es tu dia de suerte. Mueres,")
    @unusedMonsters << Monster.newMonster("El gran Cuthulu",20,pr22,bc22,4)
    
    pr23 = Prize.new(2,1)
    bc23 = NumericBadConsequence.new("Tu gobierno te recorta 2 niveles", 2, 0, 0)
    @unusedMonsters << Monster.newMonster("Serpiente Politica",8,pr23,bc23,-2)
    
    pr24 = Prize.new(1,1)
    bc24 = SpecificBadConsequence.new("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas", 0,
      [[TreasureKind::HELMET],[TreasureKind::ARMOR]],[TreasureKind::BOTHHANDS])
    @unusedMonsters << Monster.newMonster("Felpuggoth", 2,pr24,bc24,5)
    
    pr25 = Prize.new(4,2)
    bc25 = NumericBadConsequence.new("Pierdes 2 nieves", 2, 0, 0)
    @unusedMonsters << Monster.newMonster("Shoggoth",16,pr25,bc25,-4)
    
    pr26 = Prize.new(1,1)
    bc26 = NumericBadConsequence.new("Pintalabios negro. Pierdes 2 nieves", 2, 0, 0)
    @unusedMonsters << Monster.newMonster("Lolitagooth", 2, pr26, bc26,3)
  end
  
  def initCultistCardDeck
    @unusedCultists << Cultist.new("Secretario",1)
    @unusedCultists << Cultist.new("Secretario",2)
    @unusedCultists << Cultist.new("Secretario",1)
    @unusedCultists << Cultist.new("Secretario",2)
    @unusedCultists << Cultist.new("Secretario",1)
    @unusedCultists << Cultist.new("Secretario",1)
  end
  
  def initCards
    initMonstersCardDeck
    shuffleMonsters
    
    initTreasureCardDeck
    shuffleTreasures
    
    initCultistCardDeck
    shuffleCultists
  end
  
  def shuffleTreasures
    @unusedTreasures.shuffle!
  end
  
  def shuffleMonsters
    @unusedMonsters.shuffle!
  end
  
  def shuffleCultists
    @unusedCultists.shuffle!
  end
  
  def nextCultist
    if @unusedCultists.empty?
      initCultistCardDeck
      shuffleCultist
    end
    c = @unusedCultists.at(0)
    c
  end
  
  def giveTreasureBack(t)
    @usedTreasures<<t
  end
  
  def giveMonsterBack(m)
    @usedMonsters<<m
  end
  
  def nextMonster
    if @unusedMonsters.empty?   #comprobamos si tiene cartas en el mazo
      @usedMonsters.each do |m| #agregamos las cartas descartadas al mazo sin usar
        @unusedMonsters<<m
      end
      
      shuffleMonsters  #barajamos
      @usedMonsters.clear #limpiamos mazo de descartes
    end 
    m=@unusedMonsters.at(0) #cogemos la primera carta del mazo
    giveMonsterBack(m)         #la agregamos al mazo de descartes
    @unusedMonsters.delete(m) #la eliminamos del mazo
    #m                       #y devolvemos la carta
    #@unusedMonsters.at(0)    
    m
  end

  def nextTreasure
    if @unusedTreasures.empty?
      @usedTreasures.each do |t|
        @unusedTreasures<<t
      end
      
      shuffleTreasures
      @usedTreasures.clear
    end
    t=@unusedTreasures.at(0)
    @usedTreasures<<t
    @unusedTreasures.delete(t)
    t
  end
  
  def self.getInstance
   CardDealer.instance
  end
  
  
end

end

File: Programación Orientada a Objetos (PDOO)/Napakalaki(Java)/src/napakalaki/CardDealer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;
import java.util.*;
import java.util.Collections;
/**
 *
 * @author Josele
 */
public class CardDealer {
    private static final CardDealer instance=new CardDealer();
    private ArrayList <Treasure> unusedTreasures = new ArrayList();
    private ArrayList <Treasure> usedTreasures = new ArrayList();
    private ArrayList <Monster> unusedMonsters= new ArrayList();
    private ArrayList <Monster> usedMonsters= new ArrayList();
    private ArrayList <Cultist> unusedCultists= new ArrayList();
    
    
    private CardDealer(){
        
    }
    private void initTreasureCardDeck(){
        unusedTreasures.add(new Treasure("¡Śı mi amo!",4,TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Botas de investigación",3,TreasureKind.SHOES));
        unusedTreasures.add(new Treasure("Capucha de Cthulhu",3,TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("A prueba de babas",2,TreasureKind.ARMOR));
        unusedTreasures.add(new Treasure("Botas de lluvia ácida",1,TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Casco minero",2,TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Ametralladora ACME",4,TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Camiseta de la ETSIIT",1, TreasureKind.ARMOR));
        unusedTreasures.add(new Treasure("Clavo de rail ferroviario",3,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Cuchillo de sushi arcano",2,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Fez alópodo",3,TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Hacha prehistórica",2,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("El aparato del Pr. Tesla",4,TreasureKind.ARMOR));
        unusedTreasures.add(new Treasure("Gaita",4,TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Insecticida",2,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Escopeta de 3 cañones",4,TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Garabato mistico",2,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("La rebeca metálica",2,TreasureKind.ARMOR));
        unusedTreasures.add(new Treasure("Lanzallamas",4,TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Necro-comicón",1,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Necronomicón",5,TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Linterna a 2 manos",3, TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Necro-gnomicón",2,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Necrotelecom",2,TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Mazo de los antiguos",3,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Necro-playboycon",3,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Porra preternatural",2,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Shogulador",1,TreasureKind.BOTHHANDS));
        unusedTreasures.add(new Treasure("Varita de atizamiento",3,TreasureKind.ONEHAND));
        unusedTreasures.add(new Treasure("Tentácula de pega",2,TreasureKind.HELMET));
        unusedTreasures.add(new Treasure("Zapato deja-amigos",1,TreasureKind.SHOES));
    }
    private void initMonsterCardDeck(){
        BadConsequence bc1 = new SpecificBadConsequence("Pierdes tu armadura visible y otra oculta",0,
        new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
        new ArrayList(Arrays.asList(TreasureKind.ARMOR)));
        Prize pr1= new Prize(2,1);
        unusedMonsters.add(new Monster("3 Byakhees de bonanza", 8, bc1, pr1));
        
        BadConsequence bc2= new SpecificBadConsequence("Embobados con el lindo primigenio te descartas de tu casco visible",0,
        new ArrayList(Arrays.asList(TreasureKind.HELMET)),
        new ArrayList());
        Prize pr2= new Prize(1,1);
        unusedMonsters.add(new Monster("Tenochtitlan",2,bc2,pr2));
        
        BadConsequence bc3= new SpecificBadConsequence("El primordial bostezo contagioso. Pierdes el calzado visible",0,
        new ArrayList(Arrays.asList(TreasureKind.SHOES)),
        new ArrayList());
        Prize pr3= new Prize(1,1);
        unusedMonsters.add(new Monster("El sopor de Dunwich",2,bc3,pr3));
        
        BadConsequence bc4 =  new SpecificBadConsequence("Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 oculta",0,
        new ArrayList(Arrays.asList(TreasureKind.ONEHAND)),
        new ArrayList(Arrays.asList(TreasureKind.ONEHAND)));
        Prize pr4= new Prize(4,1);
        unusedMonsters.add(new Monster("Demonios de Magaluf",2,bc4,pr4));
        
        BadConsequence bc5= new SpecificBadConsequence("Pierdes todos tus tesoros visibles",0,
        new ArrayList(Arrays.asList(TreasureKind.HELMET,TreasureKind.ONEHAND,TreasureKind.BOTHHANDS,TreasureKind.ARMOR,TreasureKind.SHOES)),
        new ArrayList());
        Prize pr5= new Prize(3,1);
        unusedMonsters.add(new Monster("El gorron en el umbral",13,bc5,pr5));
        
        BadConsequence bc6= new SpecificBadConsequence("Pierdes la armadura visible",0,
        new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
        new ArrayList());
        Prize pr6= new Prize(2,1);
        unusedMonsters.add(new Monster("H.P. Munchcraft",6,bc6,pr6));
        
        BadConsequence bc7= new SpecificBadConsequence("Sientes bichos bajo la ropa. Descarta la armadura visible",0,
        new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
        new ArrayList());
        Prize pr7 = new Prize(1,1);
        unusedMonsters.add(new Monster("Necrofago",13,bc7,pr7));
        
        BadConsequence bc8 = new NumericBadConsequence("Pierdes 5 niveles y 3 tesoros visibles", 5, 3, 0);
        Prize pr8 = new Prize(3,2);
        unusedMonsters.add(new Monster("El rey de rosa", 11, bc8, pr8));
        
        BadConsequence bc9= new NumericBadConsequence("Toses los pulmones y pierdes 2 niveles",2,0,0);
        Prize pr9= new Prize(1,1);
        unusedMonsters.add(new Monster("Flecher",2,bc9,pr9));
        
        BadConsequence bc10 = new DeathBadConsequence ("Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estas muerto", true);
        Prize pr10 = new Prize(2,1);
        unusedMonsters.add(new Monster("Los hondos", 8, bc10, pr10));
        
        BadConsequence bc11 = new NumericBadConsequence("Pierdes 2 niveles y dos tesoros ocultos.", 2, 0, 2);
        Prize pr11 = new Prize(2, 1);
        unusedMonsters.add(new Monster("Semillas Cthulhu", 4, bc11, pr11));
        
        BadConsequence bc12 = new SpecificBadConsequence("Te intentas escaquear. Pierdes una mano visible.", 0,
        new ArrayList(Arrays.asList(TreasureKind.ONEHAND)),
        new ArrayList());
        Prize pr12 = new Prize(2,1);
        unusedMonsters.add(new Monster("Dameargo",1, bc12, pr12));
        
        BadConsequence bc13 = new NumericBadConsequence("Da mucho asquito. Pierdes 3 niveles", 3, 0, 0);
        Prize pr13 = new Prize(2, 1);
        unusedMonsters.add(new Monster("Pollipólipo volante", 3, bc13, pr13));
        
        BadConsequence bc14 = new DeathBadConsequence("No le hace gracia que pronuncien mal su nombre. Estas muerto", true);
        Prize pr14 = new Prize (3,1);
        unusedMonsters.add(new Monster("Yskhtihyssg-Goth", 14, bc14, pr14));
        
        BadConsequence bc15 = new DeathBadConsequence("La familia te atrapa. Estas muerto", true);
        Prize pr15 = new Prize(3, 1);
        unusedMonsters.add(new Monster("Familia feliz", 1, bc15, pr15));
             
        BadConsequence bc16 = new SpecificBadConsequence("La quinta directiva primaria te obliga a perder 2 niveles y un tesoro 2 manos visible.", 2,
        new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)),
        new ArrayList());
        Prize pr16 = new Prize(2, 1);
        unusedMonsters.add(new Monster("Roboggoth", 8, bc16, pr16));
        
        BadConsequence bc17 = new SpecificBadConsequence("Te asusta en la noche. Pierdes un casco visible.", 0,
        new ArrayList(Arrays.asList(TreasureKind.HELMET)),
        new ArrayList());
        Prize pr17 = new Prize(2, 1);
        unusedMonsters.add(new Monster("El espia sordo", 5, bc17, pr17)); 
        
        BadConsequence bc18 = new NumericBadConsequence("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles", 2, 5, 0);
        Prize pr18 = new Prize(2,1);
        unusedMonsters.add(new Monster("Tongue", 19, bc18, pr18));
    
        BadConsequence bc19 = new SpecificBadConsequence("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos", 3,
        new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS, TreasureKind.ONEHAND, TreasureKind.ONEHAND)),
        new ArrayList());
        Prize pr19 = new Prize(2,1);
        unusedMonsters.add(new Monster("Bicéfalo",21, bc19, pr19)); 
         
    //----------------------------------Monstruos con Secretarios-----------------------------------------------------------------------------
        BadConsequence bc20 = new SpecificBadConsequence("Pierdes 1 mano visible",0,new ArrayList(Arrays.asList(TreasureKind.ONEHAND)),new ArrayList());
        Prize pr20 = new Prize(3,1);
        unusedMonsters.add(new Monster("El mal indecible impronunciable",10,bc20,pr20,-2));
        
        BadConsequence bc21 = new SpecificBadConsequence("Pierdes tus tesoros visibles. Jajaja.",0,
        new ArrayList(Arrays.asList(TreasureKind.ONEHAND,TreasureKind.ONEHAND,TreasureKind.ARMOR,TreasureKind.BOTHHANDS,TreasureKind.HELMET,TreasureKind.SHOES)),new ArrayList());
        Prize pr21= new Prize(2,1);
        unusedMonsters.add(new Monster("Testigos Oculares",6,bc21,pr21,2));
        
        BadConsequence bc22 = new DeathBadConsequence("Hoy no es tu d ́ıa de suerte. Mueres.",true);
        Prize pr22 = new Prize(2,5);
        unusedMonsters.add(new Monster("El gran Cuthulu",20,bc22,pr22,4));
        
        BadConsequence bc23 = new NumericBadConsequence("Tu gobierno te recorta 2 niveles.",2,0,0);
        Prize pr23 = new Prize(2,1);
        unusedMonsters.add(new Monster("Serpiente Politica",8,bc23,pr23,-2));
        
        BadConsequence bc24 = new SpecificBadConsequence("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas.",0,
        new ArrayList(Arrays.asList(TreasureKind.HELMET,TreasureKind.ARMOR)),new ArrayList(Arrays.asList(TreasureKind.BOTHHANDS)));
        Prize pr24 = new Prize(1,1);
        unusedMonsters.add(new Monster("Felpuggoth",2,bc24,pr24,5));
        
        BadConsequence bc25 = new NumericBadConsequence("Pierdes 2 niveles",2,0,0);
        Prize pr25 = new Prize(4,2);
        unusedMonsters.add(new Monster("Shoggoth",16,bc25,pr25,-4));
        
        BadConsequence bc26 = new NumericBadConsequence("Pintalabios negro. Pierdes 2 niveles.",2,0,0);
        Prize pr26 = new Prize(1,1);
        unusedMonsters.add(new Monster("Lolitagooth",2,bc26,pr26,3));
    }
    
     private void initCultistCardDeck(){
         unusedCultists.add(new Cultist("Secretario",1));
         unusedCultists.add(new Cultist("Secretario",2));
         unusedCultists.add(new Cultist("Secretario",1));
         unusedCultists.add(new Cultist("Secretario",2));
         unusedCultists.add(new Cultist("Secretario",1));
         unusedCultists.add(new Cultist("Secretario",1));
        
    }
    
    
    
    private void shuffleTreasures(){
        Collections.shuffle(unusedTreasures);
    }
    private void shuffleMonsters(){
        Collections.shuffle(unusedMonsters);
    }
    public static CardDealer getInstance(){
        return instance;
    }
    public Treasure nextTreasure(){
        if(unusedTreasures.isEmpty()){
            for(Treasure t : usedTreasures){
                unusedTreasures.add(t);
            }
            shuffleTreasures();
            usedTreasures.clear();
        }
        Treasure T = unusedTreasures.get(0);
        usedTreasures.add(T);
        unusedTreasures.remove(T);
        return T;                 
    }
    public Monster nextMonster(){
            if(unusedMonsters.isEmpty()){
            for(Monster m : usedMonsters){
                unusedMonsters.add(m);
            }
            shuffleMonsters();
            usedMonsters.clear();
        }
        Monster M = unusedMonsters.get(0);
        usedMonsters.add(M);
        unusedMonsters.remove(0);
        return M;                 
    }
    public void giveTreasureBack(Treasure t){
        usedTreasures.add(t);
    }
    public void giveMonsterBack(Monster m){
        usedMonsters.add(m);
    }
    public void initCards(){
        this.initMonsterCardDeck();
        shuffleMonsters();
        this.initTreasureCardDeck();
        shuffleTreasures();
       
        this.initCultistCardDeck();
        shuffleCultist();
    }
    
    private void shuffleCultist(){
        Collections.shuffle(unusedCultists);
    }
   
    public Cultist nextCultist(){
        if(this.unusedCultists.isEmpty()){
            this.initCultistCardDeck();
            this.shuffleCultist();
        }
        Cultist c = this.unusedCultists.get(0);
        
        return c;
    }
}
